eth_account/_utils/structured_data/hashing.py

```diff
@@ -0,0 +1,279 @@
+    this function converts it to ``TYPE NAME``
+    """
+    return "{0} {1}".format(field["type"], field["name"])
+
+
+def encode_struct(struct_name, struct_field_types):
+    return "{0}({1})".format(
+        struct_name,
+        ','.join(map(field_identifier, struct_field_types)),
+    )
+
+
+def encode_type(primary_type, types):
+    """
+    The type of a struct is encoded as name ‖ "(" ‖ member₁ ‖ "," ‖ member₂ ‖ "," ‖ … ‖ memberₙ ")"
+    where each member is written as type ‖ " " ‖ name.
+    """
+    # Getting the dependencies and sorting them alphabetically as per EIP712
+    deps = get_dependencies(primary_type, types)
+    sorted_deps = (primary_type,) + tuple(sorted(deps))
+
+    result = ''.join(
+        [
+            encode_struct(struct_name, types[struct_name])
+            for struct_name in sorted_deps
+        ]
+    )
+    return result
+
+
+def hash_struct_type(primary_type, types):
+    return keccak(text=encode_type(primary_type, types))
+
+
+def is_valid_abi_type(type_name):
+    """
+    This function is used to make sure that the ``type_name`` is a valid ABI Type.
+
+    Please note that this is a temporary function and should be replaced by the corresponding
+    ABI function, once the following issue has been resolved.
+    https://github.com/ethereum/eth-abi/issues/125
+    """
+    valid_abi_types = {"address", "bool", "bytes", "int", "string", "uint"}
+    is_bytesN = type_name.startswith("bytes") and 1 <= int(type_name[5:]) <= 32
+    is_intN = (
+        type_name.startswith("int") and
+        8 <= int(type_name[3:]) <= 256 and
+        int(type_name[3:]) % 8 == 0
+    )
+    is_uintN = (
+        type_name.startswith("uint") and
+        8 <= int(type_name[4:]) <= 256 and
+        int(type_name[4:]) % 8 == 0
+    )
+
+    if type_name in valid_abi_types:
+        return True
+    elif is_bytesN:
+        # bytes1 to bytes32
+        return True
+    elif is_intN:
+        # int8 to int256
+        return True
+    elif is_uintN:
+        # uint8 to uint256
+        return True
+
+    return False
+
+
+def is_array_type(type):
+    # Identify if type such as "person[]" or "person[2]" is an array
+    abi_type = parse(type)
+    return abi_type.is_array
+
+
+def get_array_dimensions(data):
+    """
+    Given an array type data item, check that it is an array and
+    return the dimensions as a tuple.
+    Ex: get_array_dimensions([[1, 2, 3], [4, 5, 6]]) returns (2, 3)
+    """
+    if not isinstance(data, list) and not isinstance(data, tuple):
+        # Not checking for Iterable instance, because even Dictionaries and strings
+        # are considered as iterables, but that's not what we want the condition to be.
+        return ()
+
+    expected_dimensions = get_array_dimensions(data[0])
+    for index in range(1, len(data)):
+        # 1 dimension less sub-arrays should all have the same dimensions to be a valid array
+        if get_array_dimensions(data[index]) != expected_dimensions:
```

Comment:
I see two issues with this.  I think this is a pretty costly operation since it recursively re-computes these array dimensions making this function have something like an `O(N*2)` complexity or worse.  

Maybe you could re-structure this as something like the following and do this in two stages.

```python
@to_tuple
def enumerate_depth_and_dimensions(data, depth):
    """
    Yields 2-tuples of depth and length of element at that depth
    """
    if not is_non_string_iterable(v):
        return
    
    yield depth, len(data)

    for item in data:
        yield from enumerate_depth_and_dimensions(item, depth + 1)

def get_array_dimensions(data):
    depths_and_dimensions = get_depths_and_dimensions(data)
    # re-form as a dictionary with `depth` as key, and all of the dimensions found at that depth.
    grouped_by_depth = toolz.groupby(operator.itemgetter(0), depths_and_dimensions)
    # validate that there is only one dimension for any given depth.
    invalid_depths = tuple(depth for depth, dimensions in grouped_by_depths.items() if len(set(dimensions)) != 1)
    if invalid_depths:
        raise ValidationError("...")
    dimensions = tuple(toolz.first(set(dimensions)) for depth, dimensions in sorted(grouped_by_depth.items()))
```

This is of course untested psuedo-code and should be cleaned up for readabiltity but it should demonstrate the following algorithem.

- walk the full trie producing an enumeration of each element by it's depth and dimension.
- group all of the values together by depth and verify that each depth only has a single dimension, raising an error if this is not true.
- collate the dimension data sorted by depth

The error messaging could be improved if the enumeration of the depths and dimensions also contained the path to that part of the data which would allow for better error reporting.

Reply:
@pipermerriam I don't think there is a **difference** in the `time complexity` of my approach when compared to yours. But the approach that you suggested is easy to understand. Hence I will change it.

Reply:
Ah my bad, I see where your approach works better.